File: problems/238-product-of-array-except-self/csharp/take_skip_solution.cs
using System.Linq;
using System;

public static class Program
{
  public static void Main(string[] args)
  {
    Solution s = new Solution();
    Console.WriteLine(string.Join(", ", s.ProductExceptSelf(new int[] { 1, 2, 3, 4 })));
    Console.WriteLine(string.Join(", ", s.ProductExceptSelf(new int[] { -1, 1, 0, -3, 3 })));
  }
}

public class Solution
{
  public int[] ProductExceptSelf(int[] nums)
  {
    int[] result = new int[nums.Length];

    for (int i = 1; i <= nums.Length; i++)
    {
      int[] leftSide = nums.Take(i - 1).ToArray();
      int left = ArrayProduct(leftSide);

      int[] rightSide = nums.Skip(i).ToArray();
      int right = ArrayProduct(rightSide);
      result[i - 1] = left * right;
    }

    return result;
  }
  int ArrayProduct(int[] nums)
  {
    int p = 1;
    for (int i = 0; i < nums.Length; i++)
      p *= nums[i];
    return p;
  }
}

File: problems/238-product-of-array-except-self/python/take_skip_solution.py
from functools import reduce
from itertools import islice
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        result: List[int] = [0 for _ in nums]

        for i in range(1, len(nums) + 1):
            leftSide: List[int] = list(islice(nums, i - 1))
            left: int = reduce(lambda a, b: a * b, leftSide, 1)

            rightSide: List[int] = list(islice(nums, i, None))
            right: int = reduce(lambda a, b: a * b, rightSide, 1)

            result[i - 1] = left * right

        return result


def test(nums: List[int]) -> None:
    result: int = Solution().productExceptSelf(nums)
    print(f"Product except self for these numbers is {
          result} \n -> numbers: {nums}")


if __name__ == "__main__":
    test([1, 2, 3, 4])
    test([-1, 1, 0, -3, 3])
